### PR TITLE
klog 2.20.0, logr v1.1.0, zapr v1.1.0

### LINKED ISSUE #104097: Migrate Json logging format to use zapr
#### What would you like to be added:

/cleanup
[Structured Logging KEP](https://github.com/kubernetes/enhancements/issues/1602) proposed to introduce alternative JSON logging format, that is injected into klog by providing a [logr.Logger](https://github.com/go-logr/logr) interface implementation. 

When Json format was originally implemented we decided to write our own logr.Logger implementation to speed up work and skip taking a dependency on implementation that might not be matching specification given in KEP. 

Our implementation: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/json/json.go

As we added more testing and logr.Logger released 1.0.0, it doesn't make sense to maintain this code as it's very similar to existing logr.Logger implementation called zapr. https://github.com/go-logr/zapr

Not to person that will work on it:
We can either wait for klog and k/k to switch to logr 1.0.0 or migrate now when we are still on 0.4.0. I prefer to do it earier as it will reduce work needed to migrate to 1.0.0.

#### Why is this needed:
Reduce maintanance cost of library and consolidate work on JSON logr implementation.
It should also help when transitioning to logr 1.0.0

/cc @kubernetes/wg-structured-logging-members 
/wg structured-logging
/sig instrumentation
/priority important-soon
/cc @timhockin @pohly

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

This replaces the experimental logr v0.4 with the stable v1.1.0
release. This is a breaking API change for some users because:
  - Comparing logr.Logger against nil is not possible anymore:
    it's now a struct instead of an interface. Code which
    allows a nil logger must switch to *logr.Logger as type.
  - Logger implementations must be updated in lockstep.
    
Instead of updating the forked zapr code in json.go, directly using
the original go-logr/zapr is simpler and avoids duplication of effort.
    
#### Which issue(s) this PR fixes:
Fixes #104097

#### Special notes for your reviewer:

The updated zapr supports logging of numeric verbosity. Error messages
don't have a verbosity (= always get logged), so "v" is not getting
added to them anymore.
    
Source code logging for panic messages got fixed so that it references
the code with the invalid log call, not the json.go implementation.
    
Finally, zapr includes additional information in its panic
messages ("zap field", "ignored key", "invalid key").

#### Does this PR introduce a user-facing change?
```release-note
Users of LogFormatRegistry in component-base must update their code to use the logr v1.0.0 API. The JSON log output now uses the format from go-logr/zapr (no `v` field for error messages, additional information for invalid calls) and has some fixes (correct source code location for warnings about invalid log calls).
```

### CODE REVIEW

**Comment by pohly** [comment]
This is WIP because klog doesn't have a formal release yet and because there are known test failures in staging/src/k8s.io/component-base/logs/json.

**Comment by pohly** [comment]
This is failing because it uses `Logger{}` to denote an uninitialized logger instead of the former nil, but that then cause nil pointer access because klog does not (and cannot!) detect that such a struct is unusable.

**Comment by thockin** [comment]
Can you use `logr.Discard()` instead, if you need a valid Logger?

**Comment by pohly** [comment]
Depends on https://github.com/kubernetes/klog/pull/257

**Comment by pohly** [comment]
> Can you use `logr.Discard()` instead, if you need a valid Logger

No, klog needs to distinguish between "no logger" (= fall back to its own code) and "have logger" (= call logger). The discard logger is a valid logger and should be called if set.

On Slack we agreed to change code which needs to make that distinction to `*logr.Logger`. This is already done in klog internally. In k/k, the log format registry needs to be updated to that type.

**Comment by pohly** [comment]
At this point the only technical problem is the missing "v" field in the JSON output when using upstream zapr. This breaks the `k8s.io/kubernetes/vendor/k8s.io/component-base/logs/json` unit test. Discussion around that is in https://github.com/kubernetes/kubernetes/issues/104097

**Comment by thockin** [comment]
Reminder: I often miss @ mentions - the best way to catch my eye is to
assign to me or ping me issues directly :(

File: staging/src/k8s.io/component-base/logs/json/json_test.go [hunk 1]
```diff
@@ -147,7 +147,7 @@ func TestZapLoggerError(t *testing.T) {
 	logStr := buffer.String()
 	var ts float64
 	var lineNo int
-	expectFormat := `{"ts":%f,"caller":"json/json_test.go:%d","msg":"wrong namespace","v":0,"ns":"default","podnum":2,"time":"1µs","err":"invalid namespace:default"}`
+	expectFormat := `{"ts":%f,"caller":"json/json_test.go:%d","msg":"wrong namespace","ns":"default","podnum":2,"time":"1µs","err":"invalid namespace:default"}`
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/json/json_test.go:150]: Removal of "v" in error logs should be mentioned in changelog.

**Comment by serathius** [comment]
This looks great! Awesome job!
Aside of adding entry to changelog about removal of "v" in error logs LGTM

**Comment by pohly** [comment]
> This PR may require API review.

The "api-change" label probably was a bit too conservative: while some Go APIs change, that's not tracked via that label. The JSON log format changes slightly, but that is still alpha anyway.

**Comment by pohly** [comment]
@kubernetes/dep-approvers: @serathius finished the technical review, now we need approval for the modified dependencies (external zapr instead of internal fork, updated dependencies).

**Comment by pohly** [comment]
Just noticed some left-over comment and a formatting issue, force-pushed.

File: staging/src/k8s.io/component-base/logs/json/klog_test.go [hunk 2]
```diff
@@ -149,57 +163,63 @@ func TestKlogIntegration(t *testing.T) {
 			fun: func() {
 				klog.Error("test ", 1)
 			},
-			format: `{"ts":%f,"caller":"json/klog_test.go:%d","msg":"test 1\n","v":0}`,
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/json/klog_test.go:152]: is it expected that these no longer output `"v":0`?

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/json/klog_test.go:152]: Yes, verbosity for error didn't have any semantic meaning. It was always supposed to be equal zero, but we decided to drop it in zapr.

**Comment by liggitt** [comment]
are there any benchmarks/performance tests on the structured logging to demonstrate this replumbing didn't noticeably affect resource usage?

**Comment by liggitt** [comment]
dependency mechanics lgtm, just a couple questions

File: staging/src/k8s.io/component-base/logs/json/klog_test.go (hunk 2 above)
Comment (pohly) [reply staging/src/k8s.io/component-base/logs/json/klog_test.go:152]: Yes. Only info messages have a verbosity level and are subjected to filtering by verbosity. Error messages are always printed and therefore conceptually don't have verbosity. In other words, `V(3).Error` is the same as `V(0).Error`.

**Comment by pohly** [comment]
> are there any benchmarks/performance tests on the structured logging to demonstrate this replumbing didn't noticeably affect resource usage?

I ran the `staging/src/k8s.io/component-base/logs/json` benchmarks without and with these changes and got:
```
$ benchstat /tmp/logr04 /tmp/logr10
name               old time/op    new time/op    delta
InfoLoggerInfo-36    1.84µs ± 8%    1.93µs ± 8%     ~     (p=0.246 n=5+5)
ZapLoggerError-36    2.89µs ± 2%    2.87µs ± 4%     ~     (p=0.841 n=5+5)
ZapLoggerV-36        2.23µs ± 2%    2.14µs ± 3%   -3.80%  (p=0.016 n=5+5)

name               old alloc/op   new alloc/op   delta
InfoLoggerInfo-36    1.43kB ± 0%    1.43kB ± 0%     ~     (p=0.111 n=5+5)
ZapLoggerError-36    1.62kB ± 0%    1.56kB ± 0%   -3.93%  (p=0.008 n=5+5)
ZapLoggerV-36        1.45kB ± 0%    1.43kB ± 0%   -1.21%  (p=0.000 n=5+4)

name               old allocs/op  new allocs/op  delta
InfoLoggerInfo-36      7.00 ± 0%      7.00 ± 0%     ~     (all equal)
ZapLoggerError-36      10.0 ± 0%      10.0 ± 0%     ~     (all equal)
ZapLoggerV-36          8.00 ± 0%      7.00 ± 0%  -12.50%  (p=0.008 n=5+5)
```

So performance gets slightly better.

**Comment by serathius** [comment]
> > are there any benchmarks/performance tests on the structured logging to demonstrate this replumbing didn't noticeably affect resource usage?
> 
> I ran the `staging/src/k8s.io/component-base/logs/json` benchmarks without and with these changes and got:
> 
> ```
> $ benchstat /tmp/logr04 /tmp/logr10
> name               old time/op    new time/op    delta
> InfoLoggerInfo-36    1.84µs ± 8%    1.93µs ± 8%     ~     (p=0.246 n=5+5)
> ZapLoggerError-36    2.89µs ± 2%    2.87µs ± 4%     ~     (p=0.841 n=5+5)
> ZapLoggerV-36        2.23µs ± 2%    2.14µs ± 3%   -3.80%  (p=0.016 n=5+5)
> 
> name               old alloc/op   new alloc/op   delta
> InfoLoggerInfo-36    1.43kB ± 0%    1.43kB ± 0%     ~     (p=0.111 n=5+5)
> ZapLoggerError-36    1.62kB ± 0%    1.56kB ± 0%   -3.93%  (p=0.008 n=5+5)
> ZapLoggerV-36        1.45kB ± 0%    1.43kB ± 0%   -1.21%  (p=0.000 n=5+4)
> 
> name               old allocs/op  new allocs/op  delta
> InfoLoggerInfo-36      7.00 ± 0%      7.00 ± 0%     ~     (all equal)
> ZapLoggerError-36      10.0 ± 0%      10.0 ± 0%     ~     (all equal)
> ZapLoggerV-36          8.00 ± 0%      7.00 ± 0%  -12.50%  (p=0.008 n=5+5)
> ```
> 
> So performance gets slightly better.

Just a note about using benchstat. You should look at value of `p`. It represents statistical significance. Lower the better. Result of 0.841 for `ZapLoggerError-36` means that number of runs (5 in this example) were not enough to determine if there is improvement (variance is bigger than improvement). In this case this doesn't make much differece, but it's usually good to increase number of test runs to make sure p is kept low (<0.1).

**Comment by pohly** [comment]
/retest

Unit test flake (`k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset: TestDifferentWidths`)

**Comment by pohly** [comment]
/retest

gce test flake (`Kubernetes e2e suite: [sig-storage] Downward API volume should update labels on modification`)

**Pull request merged**